File: backend/app/models/ticket.py
from sqlalchemy import Column, Integer, String, Text, DateTime, Enum, ForeignKey, Index
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.db.session import Base
import enum


class TicketStatus(str, enum.Enum):
    OPEN = "Open"
    IN_PROGRESS = "In Progress"
    CLOSED = "Closed"


class TicketPriority(str, enum.Enum):
    LOW = "Low"
    MEDIUM = "Medium"
    HIGH = "High"
    URGENT = "Urgent"


class Ticket(Base):
    __tablename__ = "tickets"

    id = Column(Integer, primary_key=True, index=True)
    ticket_id = Column(String(20), unique=True, index=True, nullable=False)
    customer_name = Column(String(255), nullable=False)
    customer_email = Column(String(255), nullable=False, index=True)
    subject = Column(String(500), nullable=False)
    description = Column(Text, nullable=False)
    status = Column(Enum(TicketStatus), default=TicketStatus.OPEN, nullable=False, index=True)
    priority = Column(Enum(TicketPriority), default=TicketPriority.MEDIUM, nullable=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)

    notes = relationship("Note", back_populates="ticket", cascade="all, delete-orphan", order_by="Note.created_at")

    __table_args__ = (
        Index("ix_tickets_created_at", "created_at"),
        Index("ix_tickets_status_created_at", "status", "created_at"),
    )


class Note(Base):
    __tablename__ = "notes"

    id = Column(Integer, primary_key=True, index=True)
    ticket_id = Column(Integer, ForeignKey("tickets.id", ondelete="CASCADE"), nullable=False, index=True)
    note_text = Column(Text, nullable=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)

    ticket = relationship("Ticket", back_populates="notes")

File: frontend/src/types/ticket.ts
export type TicketStatus = 'Open' | 'In Progress' | 'Closed';
export type TicketPriority = 'Low' | 'Medium' | 'High' | 'Urgent';

export interface TicketListItem {
  ticket_id: string;
  customer_name: string;
  customer_email: string;
  subject: string;
  status: TicketStatus;
  priority: TicketPriority;
  created_at: string;
  updated_at: string;
}

export interface Note {
  id: number;
  note_text: string;
  created_at: string;
}

export interface TicketDetail extends TicketListItem {
  description: string;
  notes: Note[];
}

export interface CreateTicketRequest {
  customer_name: string;
  customer_email: string;
  subject: string;
  description: string;
  priority: TicketPriority;
}

export interface CreateTicketResponse {
  ticket_id: string;
  created_at: string;
}

export interface UpdateTicketRequest {
  status?: TicketStatus;
  priority?: TicketPriority;
  note?: string;
}

export interface UpdateTicketResponse {
  success: boolean;
  updated_at: string;
}

export interface TicketDeleteResponse {
  success: boolean;
}

export interface TicketStats {
  total: number;
  open: number;
  in_progress: number;
  closed: number;
}

export interface TicketsListParams {
  search?: string;
  status?: TicketStatus;
  priority?: TicketPriority;
  page?: number;
  page_size?: number;
}